### PR TITLE
fix(datasource/docker): Artifactory next link is broken for tags api

### PR DESCRIPTION
**Pull request by chezsmithy**
## Changes

* Updated getDockerApiTags to include branch logic for JFrog Artifactory when constructing the next page call to get tags when the number of docker tags is over 10,000.
* Updated the tests for Frog Artifactory to trigger this specific branch to meet code coverage requirements
* Validated the approach will not break if/when JFrog resolves the issue.

## Context

* Docker Images such as openjdk have now exceeded 10,000 tags.
* Currently when pulling the tags through JFrog tags API when using path based virtual docker repositories such as https://jrog.io/docker/openjdk the path "docker" is lost on the link portion of the response to the API calls. As such subsequent calls for the tags beyond the first 10,000 fails with a 404 causing the entire image pull to fail.
* This is documented in JFrog deferred issue [RTFACT-18971](https://jfrog.atlassian.net/browse/RTFACT-18971) and noted in code comments. JFrog has not committed a date to resolve this issue.
* Given openjdk is a common image targeted for image replacements, when attempting to replace the image against a JForg virtual docker repository containing docker hub the lookup fails causing all replacements to fail.

This has been tested against repo: https://github.com/chezsmithy/openjdk-replacement

## Documentation (please check one with an [x])

- [ ] I have updated the documentation, or
- [X] No documentation update is required

## How I've tested my work (please select one)

I have verified these changes via:

- [ ] Code inspection only, or
- [ ] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [X] Both unit tests + ran on a real repository

### CODE REVIEW

File: lib/modules/datasource/docker/common.ts [hunk 1]
```diff
@@ -10,6 +10,8 @@ export const gitRefLabel = 'org.opencontainers.image.revision';
 
 const JFROG_ARTIFACTORY_RES_HEADER = 'x-jfrog-version';
 
-export function isArtifactoryServer(res: HttpResponse | undefined): boolean {
+export function isArtifactoryServer(
+  res: HttpResponse | HttpResponse<{ tags: string[] }> | undefined
```
Comment (rarkins) [inline lib/modules/datasource/docker/common.ts:14]:
```suggestion
  res: HttpResponse<unknown> | undefined
```

**Review by rarkins — CHANGES_REQUESTED** [review]
How exactly will this "fall back" to normal behavior once it's fixed?

**Comment by chezsmithy** [comment]
@rarkins I've validated that the manually constructed next link matches what should be returned in the API call so it won't break if/when Jfrog fixes it. I could add further validation to check if the next link is actually broken by checking for the missing virtual path. Thoughts?

**Comment by rarkins** [comment]
It would be nice if this could fall back to using a fixed (correct) path once one is served, but I don't consider it mandatory to accept this PR

File: lib/modules/datasource/docker/common.ts [hunk 2]
```diff
@@ -10,6 +10,8 @@ export const gitRefLabel = 'org.opencontainers.image.revision';
 
 const JFROG_ARTIFACTORY_RES_HEADER = 'x-jfrog-version';
 
-export function isArtifactoryServer(res: HttpResponse | undefined): boolean {
+export function isArtifactoryServer(
+  res: HttpResponse<unknown> | undefined
```
Comment (viceice) [inline lib/modules/datasource/docker/common.ts:14]:
```suggestion
export function isArtifactoryServer<T = unknown>(
  res: HttpResponse<T> | undefined
```

**Review by viceice — COMMENTED** [review]
failed tests

File: lib/modules/datasource/docker/index.spec.ts [hunk 3]
```diff
@@ -1193,32 +1193,45 @@ describe('modules/datasource/docker/index', () => {
...
-          .reply(200, '', {});
-        const res = await getPkgReleases({
-          datasource: DockerDatasource.id,
-          depName: 'org.jfrog.io/virtual-mirror/node',
-        });
-        expect(res?.releases).toHaveLength(1);
-      }
-    );
+    it('jfrog artifactory - retry tags for official images by injecting `/library` after repository and before image', async () => {
+      const tags1: string[] = [...Array(10000)].map((_, i) => `${i}.0.0`);
```
Comment (viceice) [inline lib/modules/datasource/docker/index.spec.ts:1197]: i think we've a range util function for this

Comment (chezsmithy) [reply lib/modules/datasource/docker/index.spec.ts]: I see what you are referencing. I've made an update.

**Review by viceice — APPROVED** [review]

**Review by rarkins — APPROVED** [review]

**Comment by chezsmithy** [comment]
@setchy FYI on this artifactory fix for another bug.

**Pull request merged**